File: apps/engine/src/runtime/tools/knowledge/search/clubs.rs
//! Student organizations on Sun Devil Central.

use super::{Freshness, LiveSource, Param};

/// The Sun Devil Central group directory.
pub struct Clubs;

const PARAMS: &[Param] = &[Param::text("keywords", "Club name or topic.")
    .required()
    .example("artificial intelligence")];

impl LiveSource for Clubs {
    fn key(&self) -> &'static str {
        "clubs"
    }

    fn hint(&self) -> &'static str {
        "student organizations"
    }

    fn label(&self) -> &'static str {
        "Sun Devil Central"
    }

    fn category(&self) -> &'static str {
        "clubs"
    }

    fn params(&self) -> &'static [Param] {
        PARAMS
    }

    fn freshness(&self) -> Freshness {
        // Login-gated and read through the admin session, so its answers are never indexed.
        Freshness::Live
    }
}

File: apps/scraper/query/sources/clubs.py
"""Student organizations on Sun Devil Central, searched by name or topic."""

from __future__ import annotations

from scraper.core.types import QueryParam, QuerySource
from scraper.query.params import text, url
from scraper.query.sundevil_central import extract_clubs, keywords


def to_url(params: dict[str, str]) -> str:
    """The organization directory, narrowed to a keyword."""
    return url(
        "https://sundevilcentral.eoss.asu.edu/club_signup",
        [("view", "all"), ("search", keywords(text(params, "keywords")))],
    )


QUERY = QuerySource(
    key="clubs",
    description="Search ASU student organizations and clubs on Sun Devil Central.",
    params=(
        QueryParam(
            "keywords", "Club name or topic.", required=True, example="artificial intelligence"
        ),
    ),
    to_url=to_url,
    needs_js=True,
    extractor=extract_clubs,
    category="clubs",
    auth=True,
    index=False,
)
